networking_nsxv3/common/synchronization.py

```diff
@@ -90,25 +151,204 @@ def __ne__(self, other):
+    already running. If this is not the case, we will add it to our book keeping
+    here and it will be sent to a worker, wrapped in a function that will call
+    the JobRunner on job completion to update the book keeping.
+
+    If we find the job already running, the new job will be dropped, but a counter
+    for the job will be increased to mark the job for re-execution once its done,
+    in case a change to the object had occured while the worker was running.
+
+    When a worker is done with a job, it will notify us and we can either
+    remove the id from our book keeping or, if a new job was added in the meantime,
+    mark it as ready for re-execution. Note that a job will intentionally only be
+    re-executed once, independently of how often it was added while a worker was
+    already running it.
+
+    Before taking a new job from the active queue in the runner, we first check in
+    the JobRerunner for a ready job, making sure that these jobs are prioritized
+    to pick up changes quickly.
+    """
+
+    _lname_running = 'JobRerunner-running'
+    _lname_torerun = 'JobRerunner-torerun'
+
+    def __init__(self):
+        self._running = dict()
+        self._to_rerun = collections.deque()
+
+    def get_rerunnable(self) -> Runnable:
+        # Let's also use the LockManager that is used for locking
+        # in the code already. A simpler lock might do fine, but
+        # for the LockManager we know it is working.
+
+        with LockManager.get_lock(self._lname_torerun):
+            try:
+                job = self._to_rerun.popleft()
+                job.set_rescheduled()
+                LOG.info("JobRerunner (about to rerun) %s", job.get_statline())
+            except IndexError:
+                job = None
+                LOG.debug("JobRerunner had no rerunnable job")
+
+        return job
+
+    def job_done(self, job: Runnable):
+        LOG.debug("JobRerunner job_done called for %s", job)
+        with LockManager.get_lock(self._lname_running):
+            count = self._running.get(job, 0)
+
+            if count == 1:
+                del self._running[job]
+                LOG.info("JobRerunner (done, no reruns requested) %s", job.get_statline())
+            elif count > 1:
+                # we only allow exactly one job to run at a time,
+                # all jobs arriving later will increase the counter while
+                # the job is still running or they get re-queued.
+                # if a job is in the ready deque it will at some point
+                # re-appear and so we can forget about the counter.
+                LOG.info("JobRerunner (done, %d reruns requested) %s", count - 1, job.get_statline())
+                del self._running[job]
+                with LockManager.get_lock(self._lname_torerun):
+                    self._to_rerun.append(job)
+            else:
+                # prevent the error from spreading
+                del self._running[job]
```

Comment:
This could raise `KeyError` if the `job`is not in there anymore. We can still get `count == 0`, because of `self._running.get()` in line 300.